api/src/opentrons/protocol_runner/python_executor.py

```diff
@@ -0,0 +1,30 @@
+"""Python protocol executor."""
+import asyncio
+from concurrent.futures import ThreadPoolExecutor
+from functools import partial
+
+from opentrons.protocol_api_experimental import ProtocolContext
+from .python_file_reader import PythonProtocol
+
+
+class PythonExecutor:
+    """Execute a given PythonProtocol's run method with a ProtocolContext."""
+
+    def __init__(self) -> None:
+        """Initialize the executor with a thread pool.
+
+        A PythonExecutor uses its own ThreadPoolExecutor (rather than the default)
+        to avoid thread pool exhaustion from tying up protocol execution.
+        """
+        self._loop = asyncio.get_running_loop()
+        self._thread_pool = ThreadPoolExecutor(max_workers=1)
```

Comment:
This problem wasn't introduced in this PR, but I'm just noticing it. I think we leak threads by never calling self._thread_pool.shutdown().

I'm fine leaving a todo, for now.

```suggestion
        # fixme(mm, 2021-08-09): This class should be a context manager and call
        # self._thread_pool.shutdown(). Currently, I think we leak threads.
        self._thread_pool = ThreadPoolExecutor(max_workers=1)
```